### ISSUE TITLE
[no-unused-vars] More missing detections

### ISSUE THREAD

**Issue by milesj** [bradzacher/eslint-plugin-typescript#289]
**Repro**

Run the linter on https://github.com/milesj/beemo

```JSON
{
  "rules": {
    "typescript/<rule>": "<setting>"
  }
}
```

Some examples of it not working:

- Type index access: https://github.com/milesj/beemo/blob/master/packages/cli/src/CLI.ts#L58

```
11:32  warning  'DriverContext' is defined but never used    no-unused-vars
```

- When an imported type is re-assigned and exported: https://github.com/milesj/beemo/blob/master/packages/core/src/types.ts#L34 AND https://github.com/milesj/beemo/blob/master/packages/core/src/types.ts#L62

```
6:10  warning  'Tool' is defined but never used          no-unused-vars
7:10  warning  'ExecaReturns' is defined but never used  no-unused-vars
```

- Imported type used in an intersection: https://github.com/milesj/beemo/blob/master/packages/driver-babel/src/BabelDriver.ts#L46

```
  8:18  warning  'DriverArgs' is defined but never used  no-unused-vars
  9:10  warning  'BabelArgs' is defined but never used   no-unused-vars
```

- Imported namespace reference: https://github.com/milesj/beemo/blob/master/packages/driver-webpack/src/types.ts#L9

```
  6:8  warning  'webpack' is defined but never used  no-unused-vars
```

**Expected Result**

They don't fail.

**Actual Result**

They do fail.

**Additional Info**

**Versions**

| package                    | version |
| -------------------------- | ------- |
| `eslint-plugin-typescript` | `0.14.0` |
| `typescript-eslint-parser` | `21.0.2` |
| `typescript`               | `3.2.2` |

**Comment by bradzacher** [comment]
Hi @milesj, could you please try upgrading to `1.0.0-rc.3` and see if that works for these cases?

Also just a tip, when linking to files in your repo, instead of linking to master, link to a commit. The commit link's content won't change, but the link to master will change when you push.

**Comment by j-f1** [comment]
(you can hit `y` to switch to a permalink for the current file)

**Comment by SevInf** [comment]
For me, this snippet still reports errors on `1.0.0-rc.3`:

```typescript
import { ExecutionResult } from 'graphql/execution/execute';
import { UnknownObject } from '../ts-utils';

export type RigelDispatch = <DataType extends UnknownObject>(
    action: Action<DataType>
) => Promise<ExecutionResult<DataType>>;
```

```
error  'UnknownObject' is defined but never used  typescript/no-unused-vars
```
